kani_rust/mlkem.rs

```diff
@@ -0,0 +1,106 @@
+// Copyright Kani Contributors
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+//! Check that Kani correctly verifies the contracts of
+//! `montgomery_reduce` and `fqmul`.
+//
+// kani-flags: -Zfunction-contracts
+const MLKEM_Q: i32 = 3329;
+const HALF_Q: i32 = (MLKEM_Q + 1) / 2;  // 1665
+const UINT16_MAX: i32 = 65535;
+const QINV: u32 = 62209;
+
+fn cast_uint16_to_int16(x:u16) -> i16 {
+    return x as i16;
+}
+
+/*************************************************
+ * Name:        montgomery_reduce
+ *
+ * Description: Montgomery reduction; given a 32-bit integer a, computes
+ *              16-bit integer congruent to a * R^-1 mod q, where R=2^16
+ *
+ * Arguments:   - int32_t a: input integer to be reduced
+ *
+ * Returns:     integer congruent to a * R^-1 modulo q
+ *
+ *              Bounds: For any C such that |a| < q * C, the return value
+ *              has absolute value < q (C/2^16 + 1/2).
+ *
+ *              Notable special cases:
+ *              - The Montgomery multiplication of a value of absolute value
+ *                < q * C with a signed-canonical value ( < q/2 ) has
+ *                absolute value q * (0.0254 * C + 1/2).
+ *              - The Montgomery multiplication of a value of absolute value
+ *                < q * C with a value t of |t| < q has absolute value
+ *                < q * (0.0508 * C + 1/2).
+ *              - The Montgomery multiplication of a value of absolute value
+ *                < C with a value of abs < q has absolute value
+ *                < q (C/2^16 + 1/2).
+ **************************************************/
+#[kani::requires(a > -(2 * MLKEM_Q * 32768) && a <  (2 * MLKEM_Q * 32768))]
+#[kani::ensures(|y| *y as i32 > -(3 * HALF_Q) && (*y as i32) < (3 * HALF_Q))]
```

Comment:
What's going on with the syntax here? Is there not a simpler way to say "The value returned by this function"??